logstash-core/src/main/java/org/logstash/instrument/metrics/HistoricMetric.java

```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    public Map<String,Double> availableRates() {
+        if (lastRecord.get().isBefore(clock.instant().minusSeconds(10))) {
+            record();
+        }
+
+        final Map<String, Double> rates = new HashMap<>();
+
+        final Double current = rateCurrent();
+        if (current != null) {
+            rates.put("current", current);
+        }
+        final Double lastOneMinute = rateLastOneMinute();
+        if (lastOneMinute != null) {
+            rates.put("last_1_minute", lastOneMinute);
+        }
+        final Double lastFiveMinutes = rateLastFiveMinutes();
+        if (lastFiveMinutes != null) {
+            rates.put("last_5_minutes", lastFiveMinutes);
+        }
+        final Double lastFifteenMinutes = rateLastFifteenMinutes();
+        if (lastFifteenMinutes != null) {
+            rates.put("last_15_minutes", lastFifteenMinutes);
+        }
+
+        final Double lifetime = rateLifetime();
+        if (lifetime != null) {
+            rates.put("lifetime", lifetime);
+        }
+
+        return rates;
+    }
+
+    protected Double rateCurrent() {
+        return rate(quickInstant);
+    }
+
+    protected Double rateLastOneMinute() {
+        return rate(quickOneMinute);
+    }
+
+    protected Double rateLastFiveMinutes() {
+        return rate(quickFiveMinutes);
+    }
+
+    protected Double rateLastFifteenMinutes() {
+        return rate(quickFifteenMinutes);
+    }
+
+    protected Double rateLifetime() {
+        return rate(origin);
+    }
+
+    private Double rate(final AtomicReference<Node> baselineReference) {
+        return rate(baselineReference.get());
+    }
+
+    private Double rate(final Node baseline) {
+        if (baseline == null) {
+            return null;
+        } else {
+            return newest.get().getRate(baseline);
+        }
+    }
+
+    private class Node {
+        private final AtomicReference<Node> nextNode = new AtomicReference<>();
+
+        private final T numerator;
+        private final Long denominator;
+
+        private final Instant timestamp;
+
+        Node(final Instant timestamp, final T numerator, final Long denominator) {
+            this.timestamp = timestamp;
+            this.numerator = numerator;
+            this.denominator = denominator;
+        }
+
+        double getRate(final Node baseline) {
+            final double deltaNumerator = this.numerator.doubleValue() - baseline.numerator.doubleValue();
+            final long deltaDenominator = this.denominator - baseline.denominator;
+
+            return deltaNumerator / deltaDenominator;
+        }
+
+        // internal method guarantees we insert at the young end of the singly-linked list, without
+        // strict regard to `timestamp` ordering since we are tolerant to minor variations
+        // and strict ordering would require us to be doubly-linked.
+        Node seekAndPrepend(final Instant timestamp, final T numerator, final Long denominator) {
```

Comment:
I think this method is the one I am least happy about, largely because in heavy contention (which we don't expect) the stack has the opportunity to get arbitrarily deep.